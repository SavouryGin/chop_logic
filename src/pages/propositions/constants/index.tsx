import { GreekSymbol, LogicalSymbolHexCode, LogicalSymbolRawInput } from 'enums';

export * from './tab-lists';

export * from './symbols';

const constants = {
  logicalOperators: [
    LogicalSymbolRawInput.Implication,
    LogicalSymbolRawInput.Conjunction,
    LogicalSymbolRawInput.Disjunction,
    LogicalSymbolRawInput.Negation,
    LogicalSymbolRawInput.Equivalence,
  ],

  parentheses: [LogicalSymbolRawInput.OpenParenthesis, LogicalSymbolRawInput.CloseParenthesis],

  directProofsEditorTableColumns: [
    {
      field: 'step',
      title: { en: '#', ru: '№' },
    },
    {
      field: 'friendlyExpression',
      title: { en: 'Formula', ru: 'Формула' },
    },
    {
      field: 'comment',
      title: { en: 'Comment', ru: 'Коммент.' },
    },
  ],

  naturalProofsEditorTableColumns: [
    {
      field: 'step',
      title: { en: '#', ru: '№' },
    },
    {
      field: 'friendlyExpression',
      title: { en: 'Formula', ru: 'Формула' },
    },
    {
      field: 'comment',
      title: { en: 'Comment', ru: 'Коммент.' },
    },
  ],

  implicationCreationInitialValues: { firstVariable: '', secondVariable: '' },

  contradictionRealization: { firstVariable: '', secondVariable: '' },

  implicationDistributionInitialValues: { firstVariable: '', secondVariable: '', thirdVariable: '' },

  implicationCreationFormula: `${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Psi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Phi} )`,

  implicationDistributionFormula: `( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Psi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Chi} )) ${LogicalSymbolHexCode.Implication} (( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} ( ${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Chi} ))`,

  contradictionRealizationFormula: `( ${LogicalSymbolHexCode.Negation}${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} (( ${LogicalSymbolHexCode.Negation}${GreekSymbol.Phi} ${LogicalSymbolHexCode.Implication} ${LogicalSymbolHexCode.Negation}${GreekSymbol.Psi} ) ${LogicalSymbolHexCode.Implication} ${GreekSymbol.Phi} )`,
};

export default Object.freeze(constants);
