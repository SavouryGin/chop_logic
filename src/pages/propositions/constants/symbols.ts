import { LogicalSymbolHexCode, LogicalSymbolRawInput } from 'enums';
import { PropositionalSymbol } from 'types';

export const preparedSymbols: { [key in string]: PropositionalSymbol } = {
  openParenthesis: {
    input: LogicalSymbolRawInput.OpenParenthesis,
    representation: '(',
    type: 'parentheses',
    position: 0,
  },

  closeParenthesis: {
    input: LogicalSymbolRawInput.CloseParenthesis,
    representation: ')',
    type: 'parentheses',
    position: 0,
  },

  implication: {
    input: LogicalSymbolRawInput.Implication,
    representation: LogicalSymbolHexCode.Implication,
    type: 'operator',
    position: 0,
  },

  conjunction: {
    input: LogicalSymbolRawInput.Conjunction,
    representation: LogicalSymbolHexCode.Conjunction,
    type: 'operator',
    position: 0,
  },

  disjunction: {
    input: LogicalSymbolRawInput.Disjunction,
    representation: LogicalSymbolHexCode.Disjunction,
    type: 'operator',
    position: 0,
  },

  equivalence: {
    input: LogicalSymbolRawInput.Equivalence,
    representation: LogicalSymbolHexCode.Equivalence,
    type: 'operator',
    position: 0,
  },

  negation: {
    input: LogicalSymbolRawInput.Negation,
    representation: LogicalSymbolHexCode.Negation,
    type: 'operator',
    position: 0,
  },
};
