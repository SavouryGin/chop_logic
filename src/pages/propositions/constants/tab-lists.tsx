import AxiomSchemesForDirectPropositionProofs from 'pages/propositions/tabs/direct-proofs/axiom-schemes';
import DirectProofsEditor from 'pages/propositions/tabs/direct-proofs/editor';
import DirectProofsPropositionDefinitions from 'pages/propositions/tabs/direct-proofs/definitions';
import HeuristicsOfPropositions from 'pages/propositions/tabs/natural-proofs/heuristics';
import NaturalProofsEditor from 'pages/propositions/tabs/natural-proofs/editor';
import NaturalProofsPropositionDefinitions from 'pages/propositions/tabs/natural-proofs/definitions';
import React from 'react';
import RulesOfInference from 'pages/propositions/tabs/natural-proofs/rules';
import texts from 'texts/propositions/elements';
import { TabItem } from 'types';

export const propositionsDirectProofsTabs: TabItem[] = [
  {
    tabContent: <DirectProofsEditor />,
    tabTitle: texts.editor,
    tabId: 'direct-proofs-editor',
  },
  {
    tabContent: <DirectProofsPropositionDefinitions />,
    tabTitle: texts.definitions,
    tabId: 'direct-proofs-definitions',
  },
  {
    tabContent: <AxiomSchemesForDirectPropositionProofs />,
    tabTitle: texts.axioms,
    tabId: 'axiom-schemes-for-direct-proofs',
  },
];

export const propositionsNaturalProofsTabs: TabItem[] = [
  {
    tabContent: <NaturalProofsEditor />,
    tabTitle: texts.editor,
    tabId: 'natural-proofs-editor',
  },
  {
    tabContent: <NaturalProofsPropositionDefinitions />,
    tabTitle: texts.definitions,
    tabId: 'natural-proofs-definitions',
  },
  {
    tabContent: <RulesOfInference />,
    tabTitle: texts.rules,
    tabId: 'rules-of-inference',
  },
  {
    tabContent: <HeuristicsOfPropositions />,
    tabTitle: texts.heuristics,
    tabId: 'heuristics-of-propositions',
  },
];
