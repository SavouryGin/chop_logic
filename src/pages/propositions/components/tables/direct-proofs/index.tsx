import React, { memo, useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'pages/propositions/constants';
import texts from 'texts/propositions/elements';
import { dpActions as actions } from 'store/propositions/direct-proofs';
import { dpSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const DirectProofsEditorTable = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const tableData = useAppSelector(selectors.tableData);
  const language = useAppSelector(settingsSelectors.language);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.selectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='direct-proofs-editor__filler'>{texts.fillerText[language]}</div>;

  return (
    <div className='direct-proofs-editor__table'>
      <Table
        columns={constants.directProofsEditorTableColumns}
        data={tableData}
        hasCheckboxColumn
        passSelectedIds={takeSelectedIds}
        id='propositions-dp-editor-table'
      />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default memo(DirectProofsEditorTable);
