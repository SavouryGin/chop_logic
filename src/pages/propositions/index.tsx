import React from 'react';
import TabList from 'components/tab-list';
import formatClassName from 'helpers/formatters/format-class-name';
import DirectProofsEditor from './elements/direct-proofs-editor';
import NaturalProofsEditor from './elements/natural-proofs-editor';
import DefinitionsOfPropositions from './elements/definitions-of-propositions';
import HeuristicsOfPropositions from './elements/heuristics-of-propositions';
import RulesOfInference from './elements/rules-of-inference';
import { TabItem } from 'types';
import { Icon } from 'enums';
import { pageTitle, tabTitles } from 'assets/texts/propositions';
import { useAppSelector } from 'hooks';
import { settingsSelectors } from 'store/settings/selectors';

import './styles.scss';

export const propositionsTabs: TabItem[] = [
  {
    tabContent: <DirectProofsEditor />,
    tabTitle: tabTitles.direct,
    tabId: 'direct-proofs-editor',
  },
  {
    tabContent: <NaturalProofsEditor />,
    tabTitle: tabTitles.natural,
    tabId: 'natural-proofs-editor',
  },
  {
    tabContent: <DefinitionsOfPropositions />,
    tabTitle: tabTitles.definitions,
    tabId: 'definitions-of-propositions',
  },
  {
    tabContent: <HeuristicsOfPropositions />,
    tabTitle: tabTitles.heuristics,
    tabId: 'heuristics-of-propositions',
  },
  {
    tabContent: <RulesOfInference />,
    tabTitle: tabTitles.rules,
    tabId: 'rules-of-inference',
  },
];

function Propositions(): React.ReactElement {
  const language = useAppSelector(settingsSelectors.getLanguage);
  return (
    <article className='propositions'>
      <h2 className={formatClassName(['propositions__title', Icon.Propositions])}>{pageTitle[language]}</h2>
      <TabList tabs={propositionsTabs} className='propositions__tabs' />
    </article>
  );
}

export default Propositions;
