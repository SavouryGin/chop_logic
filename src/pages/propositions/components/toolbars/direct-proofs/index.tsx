import Button from 'components/controls/button';
import React, { memo } from 'react';
import { ButtonID } from 'enums';
import { propositionsDPActions as actions } from 'store/propositions/direct-proofs/slice';
import { propositionsDPSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { soundPlayer } from 'helpers/sounds';
import { useAppDispatch, useAppSelector, useIsImplicationEliminationPossible } from 'hooks';

const DirectProofsEditorToolbar = () => {
  const dispatch = useAppDispatch();
  const selectedIds = useAppSelector(selectors.getSelectedIds);
  const tableDataLength = useAppSelector(selectors.getTableDataLength);
  const isReiterationDisabled = selectedIds.length !== 1;
  const isDeleteDisabled = selectedIds.length === 0;
  const isImplicationEliminationEnabled = useIsImplicationEliminationPossible(selectedIds);
  const isReplacerDisabled = tableDataLength === 0;
  const selectedItems = useAppSelector(selectors.getSelectedTableItems);

  const deleteSteps = () => {
    dispatch(actions.deleteSteps({ isConfirmed: false }));
  };

  const reiterateStep = () => {
    dispatch(actions.reiterateStep());
  };

  const openPremise = () => {
    dispatch(actions.setUpFlag({ flag: 'isPremiseOpened', value: true }));
  };

  const openIC = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationCreationOpened', value: true }));
  };

  const openID = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationDistributionOpened', value: true }));
  };

  const openCR = () => {
    dispatch(actions.setUpFlag({ flag: 'isContradictionRealizationOpened', value: true }));
  };

  const performIE = () => {
    dispatch(actions.eliminateImplication(selectedItems));
  };

  const openReplacer = () => {
    dispatch(actions.setUpFlag({ flag: 'isReplacerFormOpened', value: true }));
  };

  return (
    <div className='direct-proofs-editor__toolbar'>
      <Button buttonId={ButtonID.Premise} sound={soundPlayer.keyboard} size='large' onClick={openPremise} />
      <Button
        buttonId={ButtonID.Reiteration}
        sound={soundPlayer.keyboard}
        size='large'
        isDisabled={isReiterationDisabled}
        onClick={reiterateStep}
      />
      <Button
        buttonId={ButtonID.Replace}
        sound={soundPlayer.keyboard}
        size='large'
        onClick={openReplacer}
        isDisabled={isReplacerDisabled}
      />
      <Button buttonId={ButtonID.Delete} sound={soundPlayer.keyboard} size='large' isDisabled={isDeleteDisabled} onClick={deleteSteps} />
      <Button buttonId={ButtonID.ImplicationCreation} sound={soundPlayer.slideClick} size='large' onClick={openIC} />
      <Button buttonId={ButtonID.ImplicationDistribution} sound={soundPlayer.slideClick} size='large' onClick={openID} />
      <Button buttonId={ButtonID.ContradictionRealization} sound={soundPlayer.slideClick} size='large' onClick={openCR} />
      <Button
        buttonId={ButtonID.ImplicationElimination}
        sound={soundPlayer.slideClick}
        isDisabled={!isImplicationEliminationEnabled}
        size='large'
        onClick={performIE}
      />
    </div>
  );
};

export default memo(DirectProofsEditorToolbar);
