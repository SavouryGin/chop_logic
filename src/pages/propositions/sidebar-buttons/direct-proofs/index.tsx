import Button from 'components/controls/button';
import React from 'react';
import { ButtonID, Icon } from 'enums';
import { propositionsDPActions as actions } from 'store/propositions/direct-proofs/slice';
import { propositionsDPSelectors as selectors } from 'store/propositions/direct-proofs/selectors';

import { useAppDispatch, useAppSelector } from 'hooks';

const PropositionsDPSidebarButtons = ({ isVisible }: { isVisible: boolean }): React.ReactElement | null => {
  const dispatch = useAppDispatch();
  const tableDataLength = useAppSelector(selectors.getTableDataLength);

  if (!isVisible) {
    return null;
  }

  // const isPDFSaveDisabled = true;
  const isExportToXMLDisabled = !tableDataLength;

  const onClickExportToXML = () => {
    dispatch(actions.exportToXML());
  };

  const openFileInputForm = () => {
    dispatch(actions.setUpFlag({ flag: 'isUserFileFormVisible', value: true }));
  };

  const onCopySteps = () => {
    console.log('Copy');
  };

  const onPasteSteps = () => {
    console.log('Paste');
  };

  return (
    <>
      {/* <li>
        <Button buttonId={ButtonID.SavePDF} icon={Icon.SavePDF} isDisabled={isPDFSaveDisabled} />
        <span className='sidebar__button-span'>Save</span>
      </li> */}
      <li>
        <Button
          buttonId={ButtonID.ExportXML}
          icon={Icon.ExportXML}
          isDisabled={isExportToXMLDisabled}
          onClick={onClickExportToXML}
          size='large'
        />
      </li>
      <li>
        <Button buttonId={ButtonID.ImportXML} icon={Icon.ImportXML} size='large' onClick={openFileInputForm} />
      </li>
      <li>
        <Button buttonId={ButtonID.CopyProof} icon={Icon.ImportXML} size='large' onClick={onCopySteps} />
      </li>
      <li>
        <Button buttonId={ButtonID.PasteProof} icon={Icon.ImportXML} size='large' onClick={onPasteSteps} />
      </li>
    </>
  );
};

export default PropositionsDPSidebarButtons;
