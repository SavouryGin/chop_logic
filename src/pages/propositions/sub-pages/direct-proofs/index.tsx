import Button from 'components/controls/button';
import PropositionsToolbar from 'pages/propositions/toolbar';
import React from 'react';
import TabList from 'components/tab-list';
import formatClass from 'helpers/formatters/format-class-name';
import texts from 'texts/propositions/elements';
import { ButtonID, Icon } from 'enums';
import { propositionsDPActions } from 'store/propositions/direct-proofs';
import { propositionsDPSelectors } from 'store/propositions/direct-proofs/selectors';
import { propositionsDirectProofsTabs } from 'pages/propositions/constants';
import { settingsSelectors } from 'store/settings/selectors';
import { soundPlayer } from 'helpers/sounds';
import { useAppDispatch, useAppSelector } from 'hooks';
import './styles.scss';

const PropositionsDirectProofs = (): React.ReactElement => {
  const language = useAppSelector(settingsSelectors.language);
  const isToolbarOpened = useAppSelector(propositionsDPSelectors.isToolbarOpened);
  const dispatch = useAppDispatch();

  const toggleToolbar = () => {
    dispatch(propositionsDPActions.setUpFlag({ flag: 'isToolbarOpened', value: !isToolbarOpened }));
  };

  const toolBar = (
    <>
      <h2
        className={formatClass(['propositions-direct-proofs__title', Icon.Propositions])}
      >{`${texts.page[language]} > ${texts.direct[language]}`}</h2>
      <Button buttonId={ButtonID.Tools} onClick={toggleToolbar} icon={Icon.Sidebar} sound={soundPlayer.keyboard} view='flat' />
    </>
  );

  return (
    <article className='propositions-direct-proofs'>
      <TabList tabs={propositionsDirectProofsTabs} className='propositions-direct-proofs__tabs' toolBar={toolBar} />
      <PropositionsToolbar isOpened={isToolbarOpened} />
    </article>
  );
};

export default PropositionsDirectProofs;
