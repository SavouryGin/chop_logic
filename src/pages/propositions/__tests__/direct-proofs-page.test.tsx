import PropositionsDirectProofs from 'pages/propositions/sub-pages/direct-proofs';
import React from 'react';
import renderWithRedux from 'helpers/test-utils/render-with-redux';
import { combineReducers } from '@reduxjs/toolkit';
import { propositionsDPInitialState, propositionsDPSlice } from 'store/propositions/direct-proofs/slice';
import { propositionsDirectProofsTabs } from 'pages/propositions/constants';
import { screen } from '@testing-library/react';
import { settingsInitialState, settingsSlice } from 'store/settings/slice';
import { titles } from 'texts/propositions';

const mockedReducer = combineReducers({
  settings: settingsSlice.reducer,
  propositionsDP: propositionsDPSlice.reducer,
});

const mockedState = {
  settings: settingsInitialState,
  propositionsDP: propositionsDPInitialState,
};

describe('PropositionsDirectProofs page:', () => {
  beforeEach(() => {
    renderWithRedux(<PropositionsDirectProofs />, mockedReducer, mockedState);
  });

  it('renders the article element', () => {
    expect(screen.getByRole('article')).toBeInTheDocument();
    expect(screen.getByRole('article')).toHaveClass('propositions-direct-proofs');
  });

  it('displays the heading', () => {
    expect(screen.getByRole('heading')).toBeInTheDocument();
    expect(screen.getByRole('heading')).toHaveTextContent(titles.page.en);
  });

  it('renders the correct number of tabs', () => {
    expect(screen.getAllByRole('tab')).toHaveLength(propositionsDirectProofsTabs.length);
  });

  it('the tabs have correct titles', () => {
    for (const item of propositionsDirectProofsTabs) {
      expect(screen.getByTestId(item.tabId)).toHaveTextContent(item.tabTitle.en);
    }
  });
});
