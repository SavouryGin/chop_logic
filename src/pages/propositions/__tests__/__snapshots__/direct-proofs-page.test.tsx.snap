// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Direct Proofs page tests: should match the snapshot 1`] = `
<DocumentFragment>
  <article
    class="propositions-direct-proofs"
  >
    <h2
      class="propositions-direct-proofs__title chop-icon__cube"
    >
      Propositions &gt; Direct Proofs
    </h2>
    <div
      class="tab-list propositions-direct-proofs__tabs"
    >
      <div
        class="tab-list__tabs"
        role="tablist"
      >
        <h3
          aria-selected="true"
          class="tab-list__tab tab-list__tab_active"
          data-testid="direct-proofs-editor"
          id="direct-proofs-editor"
          role="tab"
        >
          Editor
        </h3>
        <h3
          aria-selected="false"
          class="tab-list__tab"
          data-testid="direct-proofs-definitions"
          id="direct-proofs-definitions"
          role="tab"
        >
          Definitions
        </h3>
        <h3
          aria-selected="false"
          class="tab-list__tab"
          data-testid="axiom-schemes-for-direct-proofs"
          id="axiom-schemes-for-direct-proofs"
          role="tab"
        >
          Axiom Schemes
        </h3>
      </div>
      <section
        aria-labelledby="direct-proofs-editor"
        class="tab-list__content"
        role="tabpanel"
      >
        <div
          class="direct-proofs-editor"
        >
          <div
            class="direct-proofs-editor__table"
          >
            <table
              class="table"
            >
              <thead
                class="table__head"
              >
                <tr
                  class="table__row"
                >
                  <th
                    class="table__cell"
                  >
                    <div
                      class="checkbox-input table__checkbox checkbox-input_disabled"
                    >
                      <input
                        class="checkbox-input__default"
                        disabled=""
                        id="select_all_in_propositions-dp-editor-table"
                        name="select_all_in_propositions-dp-editor-table"
                        type="checkbox"
                      />
                      <label
                        class="label checkbox-input__label chop-icon__checkbox-unchecked"
                        data-testid="label-test-id"
                        for="select_all_in_propositions-dp-editor-table"
                      >
                        <span
                          class="label__text"
                        />
                      </label>
                    </div>
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    #
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    Formula
                  </th>
                  <th
                    class="table__cell table__cell_heading"
                  >
                    Comment
                  </th>
                </tr>
              </thead>
              <tbody
                class="table__body"
              />
            </table>
            <div
              class="direct-proofs-editor__filler"
            >
              Enter a premise or axiom to start the proof
            </div>
          </div>
          <div
            class="direct-proofs-editor__toolbar"
          >
            <button
              class="button button_large"
              id="button_id_8"
              title="Enter premise"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_8"
              >
                Premise
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_9"
              title="Reiterate proof step"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_9"
              >
                Reiteration
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_10"
              title="Replace symbol"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_10"
              >
                Replace
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_11"
              title="Delete proof step"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_11"
              >
                Delete
              </span>
            </button>
            <button
              class="button button_large"
              id="button_id_12"
              title="Implication Creation"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_12"
              >
                IC
              </span>
            </button>
            <button
              class="button button_large"
              id="button_id_13"
              title="Implication Distribution"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_13"
              >
                ID
              </span>
            </button>
            <button
              class="button button_large"
              id="button_id_14"
              title="Contradiction Realization"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_14"
              >
                CR
              </span>
            </button>
            <button
              class="button button_large button_disabled"
              disabled=""
              id="button_id_15"
              title="Implication Elimination"
              type="button"
            >
              <span
                class="button__shadow"
              />
              <span
                class="button__edge"
              />
              <span
                class="button__front"
                data-testid="button_id_15"
              >
                IE
              </span>
            </button>
          </div>
        </div>
      </section>
    </div>
  </article>
</DocumentFragment>
`;
